### DISCUSSION TITLE
Will V3 support internal test classes?

### DISCUSSION THREAD

**Discussion by voroninp** [xunit/xunit#3065] [Question] [answered]
It's a real pain with V2 that test classes must be public. 

This requirement doesn't let me define a generic base class for tests and parameterize it with an internal type.

**Comment by bradwilson** [comment, marked as answer]
This has been like this for performance reasons (private reflection being significantly more expensive than public, especially as we need to inspect every type in every assembly), so no, this will not be changing.

**Reply by voroninp** [reply]
For internal classes source generator could apply assembly level attribute:
```
[assembly: Test(TestClass = typeof(InternalTestClass))]
```

In this case you don't need to scan all types in the assembly.

**Reply by bradwilson** [reply]
Source generators are a potential future optimization, but not currently being worked on (nor committed to).
